app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java

```diff
@@ -1,83 +1,311 @@
+        }
+
+        if (dots.length > 0)
+            dots[currentPage].setTextColor(colorsActive[currentPage]);
+    }
+
+    private int getItem(int i) {
+        return mViewPager.getCurrentItem() + i;
+    }
+
+    //  viewpager change listener
+    ViewPager.OnPageChangeListener mViewPagerPageChangeListener = new ViewPager.OnPageChangeListener() {
+
+        @Override
+        public void onPageSelected(int position) {
+            addBottomDots(position);
+
+            // Change the next button text from next to finish on last slide.
+            btnNext.setText(getString((position == layouts.length - 1) ? R.string.finish : R.string.cont));
+        }
+
+        @Override
+        public void onPageScrolled(int arg0, float arg1, int arg2) {
+
+        }
+
+        @Override
+        public void onPageScrollStateChanged(int arg0) {
+
+        }
+    };
+
+    /**
+     * Making notification bar transparent
+     */
+    private void changeStatusBarColor() {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
+            Window window = getWindow();
+            window.addFlags(WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS);
+            window.setStatusBarColor(Color.TRANSPARENT);
+        }
+    }
+
+    /**
+     * View pager adapter
+     */
+    public class ViewPagerAdapter extends PagerAdapter {
+        private LayoutInflater layoutInflater;
+
+        public ViewPagerAdapter() {
+        }
+
+        @Override
+        public Object instantiateItem(ViewGroup container, int position) {
+            layoutInflater = (LayoutInflater) getSystemService(Context.LAYOUT_INFLATER_SERVICE);
+
+            View view = layoutInflater.inflate(layouts[position], container, false);
+
+            /* Slide: storage permission */
+            Button btnGrantStoragePerm = (Button) view.findViewById(R.id.btnGrantStoragePerm);
+            if (btnGrantStoragePerm != null) {
+                btnGrantStoragePerm.setOnClickListener(new View.OnClickListener() {
+                    @Override
+                    public void onClick(View v) {
+                        requestStoragePermission();
+                    }
+                });
+            }
+
+            /* Slide: location permission */
+            Button btnGrantLocationPerm = (Button) view.findViewById(R.id.btnGrantLocationPerm);
+            if (btnGrantLocationPerm != null) {
+                btnGrantLocationPerm.setOnClickListener(new View.OnClickListener() {
+                    @Override
+                    public void onClick(View v) {
+                        requestLocationPermission();
+                    }
+                });
+            }
+
+            container.addView(view);
+            return view;
+        }
+
+        @Override
+        public int getCount() {
+            return layouts.length;
+        }
+
+        @Override
+        public boolean isViewFromObject(View view, Object obj) {
+            return view == obj;
+        }
```

Comment:
Why do we need to override this?

Reply:
The method  instantiateItem(ViewGroup, int) returns an Object for a particular view. PagerAdapter implementation is considering this Object as a key value when viewpager changes a page. If we return the view itself from instantiateItem(ViewGroup, int), then our key for that page becomes the view itself. We can check return view == obj; from isViewFromObject which will always return true and our pages will display.